Replace pointer and vlen with SGVector in QDA.

sudk1896 opened the pull request:
Replacing the old pointer and vlen with `SGVector`.
#3218.
@karlnapf : The tests have passed. Review.

karlnapf commented on src/shogun/multiclass/QDA.cpp:
```diff
@@ -129,21 +129,21 @@ CMulticlassLabels* CQDA::apply_multiclass(CFeatures* data)
 	int32_t vlen;
 	bool vfree;
-	float64_t* vec;
+	SGVector< float64_t > vec;
```
no space in the template please

>

karlnapf commented on src/shogun/multiclass/QDA.cpp:
```diff
@@ -129,21 +129,21 @@ CMulticlassLabels* CQDA::apply_multiclass(CFeatures* data)
 
 	int32_t vlen;
 	bool vfree;
```
this variable is not used anymore I guess?

karlnapf commented on src/shogun/multiclass/QDA.cpp:
```diff
@@ -129,21 +129,21 @@ CMulticlassLabels* CQDA::apply_multiclass(CFeatures* data)
 		{
-			vec = rf->get_feature_vector(i, vlen, vfree);
+			vec = rf->get_feature_vector(i);
 			ASSERT(vec)
```
this is not needed anymore, or needs to be modified to check vec.vec

karlnapf commented on src/shogun/multiclass/QDA.cpp:
```diff
@@ -255,21 +255,21 @@ bool CQDA::train_machine(CFeatures* data)
 	int32_t vlen;
 	bool vfree;
-	float64_t* vec;
+	SGVector< float64_t > vec;
```
as above

karlnapf commented on src/shogun/multiclass/QDA.cpp:
```diff
@@ -255,21 +255,21 @@ bool CQDA::train_machine(CFeatures* data)
 		{
-			vec = rf->get_feature_vector(class_idxs[k*num_vec + i], vlen, vfree);
+			vec = rf->get_feature_vector(class_idxs[k*num_vec + i]);
 			ASSERT(vec)
```
as above

karlnapf commented:
Thanks for that!
Can you please run the QDA unit tests with valgrind and make sure there is no leak or other error introduced?
Something like `valgrind ./shogun-unit-tests --gtest_filter=*QDA*` in `build/tests/unit`
Google for valgrind usage

karlnapf commented:
Once that's all fine, and all cosmetic issues have been addressed, please merge all commits into a single one. Then good to merge

sudk1896 commented:
@karlnapf: I think the `ASSERT(vec)` statements are fine. I removed the unused variables and implemented the style fixes. Let me know if these are okay and then i will squash them into a single commit.
The valgrind test results - 
`
sudarshan@sudarshan-HP-15-Notebook-PC:~/shogun/build/tests/unit$ valgrind ./shogun-unit-test --gtest_filter=*QDA* --leak-check=full
==4400== Memcheck, a memory error detector
==4400== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==4400== Using Valgrind-3.10.1 and LibVEX; rerun with -h for copyright info
==4400== Command: ./shogun-unit-test --gtest_filter=*QDA* --leak-check=full
==4400==
Note: Google Test filter = *QDA*
[==========] Running 5 tests from 5 test cases.
[----------] Global test environment set-up.
[----------] 1 test from SGObject
[ RUN      ] SGObject.clone_equals_QDA
[       OK ] SGObject.clone_equals_QDA (53 ms)
[----------] 1 test from SGObject (60 ms total)
 
[----------] 1 test from QDA
[ RUN      ] QDA.train_and_apply
[       OK ] QDA.train_and_apply (177 ms)
[----------] 1 test from QDA (178 ms total)
 
[----------] 1 test from DynamicObjectArray
[ RUN      ] DynamicObjectArray.array_QDA_clone_equals
[       OK ] DynamicObjectArray.array_QDA_clone_equals (9 ms)
[----------] 1 test from DynamicObjectArray (10 ms total)
 
[----------] 1 test from SerializationAscii
[ RUN      ] SerializationAscii.QDA
[       OK ] SerializationAscii.QDA (157 ms)
[----------] 1 test from SerializationAscii (157 ms total)
 
[----------] 1 test from SerializationXML
[ RUN      ] SerializationXML.QDA
[       OK ] SerializationXML.QDA (249 ms)
[----------] 1 test from SerializationXML (249 ms total)
 
[----------] Global test environment tear-down
[==========] 5 tests from 5 test cases ran. (704 ms total)
[  PASSED  ] 5 tests.
==4400==
==4400== HEAP SUMMARY:
==4400==     in use at exit: 180 bytes in 3 blocks
==4400==   total heap usage: 69,168 allocs, 69,165 frees, 4,523,448 bytes allocated
==4400==
==4400== LEAK SUMMARY:
==4400==    definitely lost: 0 bytes in 0 blocks
==4400==    indirectly lost: 0 bytes in 0 blocks
==4400==      possibly lost: 0 bytes in 0 blocks
==4400==    still reachable: 180 bytes in 3 blocks
==4400==         suppressed: 0 bytes in 0 blocks
==4400== Rerun with --leak-check=full to see details of leaked memory
==4400==
==4400== For counts of detected and suppressed errors, rerun with: -v
==4400== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)`

karlnapf commented:
ASSERT compares the element to 0, which is not what you want here, you want to check that the underlying memory is not NULL. So (as I wrote!!!) it should be `ASSERT(vec.vector)`, since `vector` member is the pointer to the data.

karlnapf commented:
I agree, the valgrind trick should be in the testing readme. I will update the readmes soon (hopefully)

karlnapf commented:
Now for the leaks you have, were they there before the changes you made? Please check

karlnapf commented:
Changes are good otherwise. Thanks!

karlnapf commented:
Changes are good otherwise. Thanks!

karlnapf commented:
While you are cleaning this QDA mess up, another thing that should go is any method called from SGVector, e.g.

```
SGVector<float64_t>::vector_multiply(col, col, col, m_dim);
SGVector<float64_t>::scale_vector(1.0/(class_nums[k]-1), col, m_dim);
```

These should be done with eigen3, then you can also avoid storing the vector pointer above. Feel free to add (but first check the memory thing above, as it is harder to trace with more changes)

sudk1896 commented:
@karlnapf : There are no leaks as you can see the valgrind output. I will incorporate your changes.

sudk1896 commented:
@karlnapf: This is the memory leak details before I made the changes-]
```
sudarshan@sudarshan-HP-15-Notebook-PC:~/shogun/build/tests/unit$ valgrind --leak-check=full --show-leak-kinds=all ./shogun-unit-test --gtest_filter=*QDA*
==3222== Memcheck, a memory error detector
==3222== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==3222== Using Valgrind-3.10.1 and LibVEX; rerun with -h for copyright info
==3222== Command: ./shogun-unit-test --gtest_filter=*QDA*
==3222== 
Note: Google Test filter = *QDA*
[==========] Running 5 tests from 5 test cases.
[----------] Global test environment set-up.
[----------] 1 test from SGObject
[ RUN      ] SGObject.clone_equals_QDA
[       OK ] SGObject.clone_equals_QDA (55 ms)
[----------] 1 test from SGObject (62 ms total)

[----------] 1 test from QDA
[ RUN      ] QDA.train_and_apply
[       OK ] QDA.train_and_apply (177 ms)
[----------] 1 test from QDA (177 ms total)

[----------] 1 test from DynamicObjectArray
[ RUN      ] DynamicObjectArray.array_QDA_clone_equals
[       OK ] DynamicObjectArray.array_QDA_clone_equals (9 ms)
[----------] 1 test from DynamicObjectArray (9 ms total)

[----------] 1 test from SerializationAscii
[ RUN      ] SerializationAscii.QDA
[       OK ] SerializationAscii.QDA (153 ms)
[----------] 1 test from SerializationAscii (153 ms total)

[----------] 1 test from SerializationXML
[ RUN      ] SerializationXML.QDA
[       OK ] SerializationXML.QDA (249 ms)
[----------] 1 test from SerializationXML (249 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 5 test cases ran. (702 ms total)
[  PASSED  ] 5 tests.
==3222== 
==3222== HEAP SUMMARY:
==3222==     in use at exit: 180 bytes in 3 blocks
==3222==   total heap usage: 69,097 allocs, 69,094 frees, 4,518,322 bytes allocated
==3222== 
==3222== 32 bytes in 1 blocks are still reachable in loss record 1 of 3
==3222==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3222==    by 0x8F5568F: _dlerror_run (dlerror.c:141)
==3222==    by 0x8F550C0: dlopen@@GLIBC_2.2.5 (dlopen.c:87)
==3222==    by 0x9FD3475: ??? (in /lib/x86_64-linux-gnu/liblzma.so.5.0.0)
==3222==    by 0x9FD7DC7: lzma_auto_decoder (in /lib/x86_64-linux-gnu/liblzma.so.5.0.0)
==3222==    by 0x88E19CC: ??? (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3222==    by 0x88E233F: __libxml2_xzread (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3222==    by 0x882FE28: ??? (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3222==    by 0x8831994: xmlParserInputBufferGrow (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3222==    by 0x8803EA4: xmlParserInputGrow (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3222==    by 0x88093C0: ??? (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3222==    by 0x881B34C: xmlParseDocument (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3222== 
==3222== 44 bytes in 1 blocks are still reachable in loss record 2 of 3
==3222==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3222==    by 0x400FD90: _dl_signal_error (dl-error.c:90)
==3222==    by 0x401468F: _dl_open (dl-open.c:715)
==3222==    by 0x8F5502A: dlopen_doit (dlopen.c:66)
==3222==    by 0x400FFC3: _dl_catch_error (dl-error.c:187)
==3222==    by 0x8F5562C: _dlerror_run (dlerror.c:163)
==3222==    by 0x8F550C0: dlopen@@GLIBC_2.2.5 (dlopen.c:87)
==3222==    by 0x9FD3475: ??? (in /lib/x86_64-linux-gnu/liblzma.so.5.0.0)
==3222==    by 0x9FD7DC7: lzma_auto_decoder (in /lib/x86_64-linux-gnu/liblzma.so.5.0.0)
==3222==    by 0x88E19CC: ??? (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3222==    by 0x88E233F: __libxml2_xzread (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3222==    by 0x882FE28: ??? (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3222== 
==3222== 104 bytes in 1 blocks are still reachable in loss record 3 of 3
==3222==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3222==    by 0x740B7A8: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==3222==    by 0x740FD8A: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==3222==    by 0x740B9CC: omp_set_dynamic (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==3222==    by 0x56C9774: shogun::Parallel::Parallel() (in /home/sudarshan/shogun/build/src/shogun/libshogun.so.17.2)
==3222==    by 0x56C95F6: shogun::init_shogun(void (*)(_IO_FILE*, char const*), void (*)(_IO_FILE*, char const*), void (*)(_IO_FILE*, char const*), void (*)(bool&, bool&)) (in /home/sudarshan/shogun/build/src/shogun/libshogun.so.17.2)
==3222==    by 0xCA4231: main (in /home/sudarshan/shogun/build/tests/unit/shogun-unit-test)
==3222== 
==3222== LEAK SUMMARY:
==3222==    definitely lost: 0 bytes in 0 blocks
==3222==    indirectly lost: 0 bytes in 0 blocks
==3222==      possibly lost: 0 bytes in 0 blocks
==3222==    still reachable: 180 bytes in 3 blocks
==3222==         suppressed: 0 bytes in 0 blocks
==3222== 
==3222== For counts of detected and suppressed errors, rerun with: -v
==3222== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

sudk1896 commented:
This is the output after I made the changes:-
```
sudarshan@sudarshan-HP-15-Notebook-PC:~/shogun/build/tests/unit$ valgrind --leak-check=full --show-leak-kinds=all ./shogun-unit-test --gtest_filter=*QDA*
==3341== Memcheck, a memory error detector
==3341== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==3341== Using Valgrind-3.10.1 and LibVEX; rerun with -h for copyright info
==3341== Command: ./shogun-unit-test --gtest_filter=*QDA*
==3341== 
Note: Google Test filter = *QDA*
[==========] Running 5 tests from 5 test cases.
[----------] Global test environment set-up.
[----------] 1 test from SGObject
[ RUN      ] SGObject.clone_equals_QDA
[       OK ] SGObject.clone_equals_QDA (56 ms)
[----------] 1 test from SGObject (61 ms total)

[----------] 1 test from QDA
[ RUN      ] QDA.train_and_apply
[       OK ] QDA.train_and_apply (178 ms)
[----------] 1 test from QDA (178 ms total)

[----------] 1 test from DynamicObjectArray
[ RUN      ] DynamicObjectArray.array_QDA_clone_equals
[       OK ] DynamicObjectArray.array_QDA_clone_equals (9 ms)
[----------] 1 test from DynamicObjectArray (10 ms total)

[----------] 1 test from SerializationAscii
[ RUN      ] SerializationAscii.QDA
[       OK ] SerializationAscii.QDA (153 ms)
[----------] 1 test from SerializationAscii (153 ms total)

[----------] 1 test from SerializationXML
[ RUN      ] SerializationXML.QDA
[       OK ] SerializationXML.QDA (249 ms)
[----------] 1 test from SerializationXML (251 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 5 test cases ran. (704 ms total)
[  PASSED  ] 5 tests.
==3341== 
==3341== HEAP SUMMARY:
==3341==     in use at exit: 180 bytes in 3 blocks
==3341==   total heap usage: 69,097 allocs, 69,094 frees, 4,518,325 bytes allocated
==3341== 
==3341== 32 bytes in 1 blocks are still reachable in loss record 1 of 3
==3341==    at 0x4C2CC70: calloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3341==    by 0x8F5568F: _dlerror_run (dlerror.c:141)
==3341==    by 0x8F550C0: dlopen@@GLIBC_2.2.5 (dlopen.c:87)
==3341==    by 0x9FD3475: ??? (in /lib/x86_64-linux-gnu/liblzma.so.5.0.0)
==3341==    by 0x9FD7DC7: lzma_auto_decoder (in /lib/x86_64-linux-gnu/liblzma.so.5.0.0)
==3341==    by 0x88E19CC: ??? (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3341==    by 0x88E233F: __libxml2_xzread (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3341==    by 0x882FE28: ??? (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3341==    by 0x8831994: xmlParserInputBufferGrow (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3341==    by 0x8803EA4: xmlParserInputGrow (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3341==    by 0x88093C0: ??? (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3341==    by 0x881B34C: xmlParseDocument (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3341== 
==3341== 44 bytes in 1 blocks are still reachable in loss record 2 of 3
==3341==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3341==    by 0x400FD90: _dl_signal_error (dl-error.c:90)
==3341==    by 0x401468F: _dl_open (dl-open.c:715)
==3341==    by 0x8F5502A: dlopen_doit (dlopen.c:66)
==3341==    by 0x400FFC3: _dl_catch_error (dl-error.c:187)
==3341==    by 0x8F5562C: _dlerror_run (dlerror.c:163)
==3341==    by 0x8F550C0: dlopen@@GLIBC_2.2.5 (dlopen.c:87)
==3341==    by 0x9FD3475: ??? (in /lib/x86_64-linux-gnu/liblzma.so.5.0.0)
==3341==    by 0x9FD7DC7: lzma_auto_decoder (in /lib/x86_64-linux-gnu/liblzma.so.5.0.0)
==3341==    by 0x88E19CC: ??? (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3341==    by 0x88E233F: __libxml2_xzread (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3341==    by 0x882FE28: ??? (in /usr/lib/x86_64-linux-gnu/libxml2.so.2.9.1)
==3341== 
==3341== 104 bytes in 1 blocks are still reachable in loss record 3 of 3
==3341==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==3341==    by 0x740B7A8: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==3341==    by 0x740FD8A: ??? (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==3341==    by 0x740B9CC: omp_set_dynamic (in /usr/lib/x86_64-linux-gnu/libgomp.so.1.0.0)
==3341==    by 0x56C9774: shogun::Parallel::Parallel() (in /home/sudarshan/shogun/build/src/shogun/libshogun.so.17.2)
==3341==    by 0x56C95F6: shogun::init_shogun(void (*)(_IO_FILE*, char const*), void (*)(_IO_FILE*, char const*), void (*)(_IO_FILE*, char const*), void (*)(bool&, bool&)) (in /home/sudarshan/shogun/build/src/shogun/libshogun.so.17.2)
==3341==    by 0xCA4231: main (in /home/sudarshan/shogun/build/tests/unit/shogun-unit-test)
==3341== 
==3341== LEAK SUMMARY:
==3341==    definitely lost: 0 bytes in 0 blocks
==3341==    indirectly lost: 0 bytes in 0 blocks
==3341==      possibly lost: 0 bytes in 0 blocks
==3341==    still reachable: 180 bytes in 3 blocks
==3341==         suppressed: 0 bytes in 0 blocks
==3341== 
==3341== For counts of detected and suppressed errors, rerun with: -v
==3341== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

sudk1896 commented:
@karlnapf : As you can see from the output, the memory leaks were already there and my changes haven't introduced any new memory leaks.
The valgrind outputs are identical. The only caveat here is well there is a memory leak, but maybe that can be dealt with in another PR.

sudk1896 commented:
About your other suggestion, it is better If I make that in another PR. Don't want too many changes in one. Let me know if this is good to merge, I will squash them into a single commit then.

karlnapf commented:
Yep I agree with you. Squash and we can merge. Thanks for the memory check!

sudk1896 commented:
@karlnapf: Squashed all the commits into one. Should be ready now.

karlnapf commented:
Cool! I will merge once travis is green. Ping me when that happens

BTW can you check which of the 5 tests is leaking memory, and why, and then file an issue for that? And then fix it? That's a very useful thing to do and lots to learn as well :)

sudk1896 commented:
@karlnapf: Sure will do.

karlnapf commented:
We cannot merge with this commit message
